net: Add CNode::ConnectedThroughNetwork member function

hebasto opened the pull request:
This PR:
- adds `CNode::ConnectedThroughNetwork` member function
- is based on #19991, and only last two commits belong to it
- is required for https://github.com/bitcoin-core/gui/pull/86 and #20002

practicalswift commented:
Concept ACK

Nice! :)

laanwj commented:
Concept ACK

What about a multiple-choice option that more generally specifies the network that the connection (comes in/goes out) through, instead of specializing it for Tor?

Also this should be 'onion' to be consistent with our other uses.

hebasto commented:
@laanwj 
> What about a multiple-choice option that more generally specifies the network that the connection (comes in/goes out) through, instead of specializing it for Tor? Might want to have the same for I2P later, for example.

Do I understand you correctly that smth following is expected:
```
$ src/bitcoin-cli help getpeerinfo | grep network_type
    "network_type" : clearnet|onion|I2P,        Type of the network the peer has connected via
```
?

naumenkogs commented:
Concept ACK.
I agree with wladimir's suggestions.
@hebasto I think your understanding is correct.

jonatack commented:
> Concept ACK
> 
> What about a multiple-choice option that more generally specifies the network that the connection (comes in/goes out) through, instead of specializing it for Tor? Might want to have the same for I2P later, for example.
> 
> Also this should be 'onion' to be consistent with our other uses.

This was the goal in #20002, which exposes the `GetNetClass()` index (allowing for a significant code simplification in -netinfo). Feedback welcome if it should also expose the network class name.

hebasto commented:
@jonatack 
> This was the goal in #20002, which exposes the `GetNetClass()` index (allowing for a significant code simplification in -netinfo). Feedback welcome if it should also expose the network class name.

What about limiting this PR to introducing `CNode::GetNetClass()` in addition to `CNetAddr::GetNetClass()` as the latter does not work for inbound connections via Tor?

jonatack commented:
@hebasto after chatting offline, that makes sense, will look at building on your commit "net: Add CNode::ConnectedViaTor() member function".

hebasto commented:
Updated fcd938e80abe634ccd0c0a289a8bdd339c186d24 -> 911a3f9929e2c10493c1ad525ad592dd6859df3f ([pr19998.02](https://github.com/hebasto/bitcoin/commits/pr19998.02) -> [pr19998.03](https://github.com/hebasto/bitcoin/commits/pr19998.03), [compare](https://github.com/hebasto/bitcoin/compare/pr19998.02..pr19998.03)):

- dropped the last commit "rpc: Add `via_tor` to `getpeerinfo` output" in favor of more consistent approach in #20002

jonatack commented on src/net.cpp:
```diff
@@ -514,6 +514,15 @@ void CNode::SetAddrLocal(const CService& addrLocalIn) {
 
+bool CNode::ConnectedViaTor() const
+{
+    if (m_conn_type == ConnectionType::INBOUND) {
```
```suggestion
    if (IsInboundConn())
```

jonatack replied:
(I'm testing building on this function ATM with the following in `CNodeStats`)
```cpp
stats.m_network = ConnectedViaTor() ? uint32_t(NET_ONION) : addr.GetNetClass();
```

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19998#issuecomment-701533052).

laanwj commented:
Haven't reveiewed the code, but conceptually, for future-readyness I'd prefer this to be called `ConnectedThroughNetwork` and return an enumeration instead of be Tor-specific.

jonatack commented:
Yes, ATM returning an enum is delegated to the follow-up #20002, where the `Network` enum is returned using the existing `GetNetworkName()` function for the enum-to-string conversion. New networks like I2P and CJDNS can be added to that function when they are supported.

hebasto committed:
net: Add CNode::m_inbound_onion data member

hebasto committed:
net: Add CNode::ConnectedThroughNetwork member function

hebasto commented:
Updated 911a3f9929e2c10493c1ad525ad592dd6859df3f -> 1cf469b7a3dd2d02e1dacc0006d631a10e45ec47 ([pr19998.03](https://github.com/hebasto/bitcoin/commits/pr19998.03) -> [pr19998.05](https://github.com/hebasto/bitcoin/commits/pr19998.05)):

- rebased on top of the recent #19991

- addressed @laanwj's [comment](https://github.com/bitcoin/bitcoin/pull/19998#issuecomment-701261700):

> Haven't reveiewed the code, but conceptually, for future-readyness I'd prefer this to be called `ConnectedThroughNetwork` and return an enumeration instead of have a Tor-specific call one the API.
> 
> Edit: oh, I suggested this before. Apparently I feel strongly about this smile .
> 
> > $ src/bitcoin-cli help getpeerinfo | grep network_type
> > "network_type" : clearnet|onion|I2P,        Type of the network the peer has connected via
> 
> Yes, exactly! Or maybe even split clearnet into IPv4 and IPv6.

hebasto commented:
Updated 911a3f9929e2c10493c1ad525ad592dd6859df3f -> cf532ae5d68c2e35e73c793923398e01ee7bca7f ([pr19998.03](https://github.com/hebasto/bitcoin/commits/pr19998.03) -> [pr19998.06](https://github.com/hebasto/bitcoin/commits/pr19998.06)):

- rebased on top of the recent #19991

- addressed @laanwj's [comment](https://github.com/bitcoin/bitcoin/pull/19998#issuecomment-701261700):

> Haven't reveiewed the code, but conceptually, for future-readyness I'd prefer this to be called `ConnectedThroughNetwork` and return an enumeration instead of have a Tor-specific call one the API.
> 
> Edit: oh, I suggested this before. Apparently I feel strongly about this smile .
> 
> > $ src/bitcoin-cli help getpeerinfo | grep network_type
> > "network_type" : clearnet|onion|I2P,        Type of the network the peer has connected via
> 
> Yes, exactly! Or maybe even split clearnet into IPv4 and IPv6.

hebasto commented:
Updated cf532ae5d68c2e35e73c793923398e01ee7bca7f -> c972792511f764ff300602bd3c71df1fea90ddcc ([pr19998.06](https://github.com/hebasto/bitcoin/commits/pr19998.06) -> [pr19998.07](https://github.com/hebasto/bitcoin/commits/pr19998.07)) to the recent #19991.

Sjors commented:
utACK c972792

This will be easier to test after #20002 is refreshed (don't forget to mention it in the description) 

The asmap test failure on macOS Travis seems spurious (it passes on my machine)

hebasto commented:
Updated c972792511f764ff300602bd3c71df1fea90ddcc -> ac0f4975fe00f2f72c408cbd6b642b29b0e2f8ce ([pr19998.07](https://github.com/hebasto/bitcoin/commits/pr19998.07) -> [pr19998.08](https://github.com/hebasto/bitcoin/commits/pr19998.08)) due to the conflict with #19991.

jonatack commented:
Do you plan to add test coverage for the changes here (and in #19991?)

hebasto commented:
@jonatack 
> Do you plan to add test coverage for the changes here (and in #19991?)

I thought [your tests](https://github.com/bitcoin/bitcoin/pull/19991#issuecomment-702685460) could cover all aspects, right?

jonatack commented:
@hebasto I don't plan to open a PR just for follow-up tests. I think it's easiest when the coverage accompanies the changes.

jonatack commented:
Starting to see some inbound onions while testing this on an updated #20002 (will push an update soon)
```text
<-> relay   net mping   ping send recv  txn  blk uptime  asmap  id address
 in  full onion   358    469    0   90               49         35 127.0.0.1:43064
 in block onion   450    722   29  104               35         66 127.0.0.1:45118
```

jonatack commented:
Starting to see some inbound onions while testing this on an updated #20002 (will push an update soon)
```text
<-> relay   net mping   ping send recv  txn  blk uptime  asmap  id address
 in  full onion   358    469    0   90               49         35 127.0.0.1:43064
 in block onion   450    722   29  104               35         66 127.0.0.1:45118
```

The updated function here is easy to use, too.
```c++
stats.m_network = GetNetworkName(ConnectedThroughNetwork());
```

jonatack commented on src/net.h:
```diff
@@ -586,6 +587,9 @@ class CConnman
 
     std::atomic<int64_t> m_next_send_inv_to_incoming{0};
 
+    //! -bind=<address>:<port>=onion arguments.
```
- this documentation is good to provide but a bit succinct; maybe also describe what `m_onion_binds` is
- could use the `/**` doxygen formatting like `m_try_another_outbound_peer` just above

hebasto replied:
[Developer Notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-doxygen-compatible-comments) suggest `//!` style to describe a member.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19998#issuecomment-703090033).

hebasto committed:
test: Add tests for CNode::ConnectedThroughNetwork

jonatack commented on src/net.cpp:
```diff
@@ -538,6 +539,14 @@ void CNode::SetAddrLocal(const CService& addrLocalIn) {
+    if (IsInboundConn() && m_inbound_onion) {
+        return NET_ONION;
+    }
+    return addr.GetNetClass();
```
optional style nit, could just simplify to a straight ternary
```diff
-    if (IsInboundConn() && m_inbound_onion) {
-        return NET_ONION;
-    }
-    return addr.GetNetClass();
+    return IsInboundConn() && m_inbound_onion ? NET_ONION : addr.GetNetClass();
```
or a one-line guard conditional
```diff
-    if (IsInboundConn() && m_inbound_onion) {
-        return NET_ONION;
-    }
+    if (IsInboundConn() && m_inbound_onion) return NET_ONION;
     return addr.GetNetClass();
```

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19998#issuecomment-703090033).

jonatack commented on src/net.h:
```diff
@@ -954,6 +958,8 @@ class CNode
         assert(false);
     }
 
+    Network ConnectedThroughNetwork() const;
```
Can you add a doxygen comment here? Maybe provide a quick context on the difference between this and `GetNetClass()` or why it's needed.

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/19998#issuecomment-703090033).

hebasto commented:
Added test coverage.

jonatack reviewed (commented):
ACK, I've been testing this with #20002 and also rebased on #19954 (it's thankfully a clean merge) and it seems to be working. A few optional comments below. It may be good to update `fuzz/netaddress.cpp` (or other fuzzers, I didn't look further) with `ConnectedThroughNetwork()` and add unit or functional tests for the new `-bind` functionality. I did add some testing in #20002 for the values returned by `ConnectedThroughNetwork()`. Edit: reviewing the new push!

jonatack commented on src/test/net_tests.cpp:
```diff
@@ -192,14 +192,34 @@ BOOST_AUTO_TEST_CASE(cnode_simple_test)
+    BOOST_CHECK_EQUAL(pnode1->ConnectedThroughNetwork(), Network::NET_IPV4);
 
-    std::unique_ptr<CNode> pnode2 = MakeUnique<CNode>(id++, NODE_NETWORK, height, hSocket, addr, 1, 1, CAddress(), pszDest, ConnectionType::INBOUND);
+    std::unique_ptr<CNode> pnode2 = MakeUnique<CNode>(id++, NODE_NETWORK, height, hSocket, addr, 1, 1, CAddress(), pszDest, ConnectionType::INBOUND, false);
```
here and lines 206 and 215
```suggestion
    std::unique_ptr<CNode> pnode2 = MakeUnique<CNode>(id++, NODE_NETWORK, height, hSocket, addr, 1, 1, CAddress(), pszDest, ConnectionType::INBOUND, /* inbound_onion= */ false);
```
could optionally also add `/*nKeyedNetGroupIn=*/ 1` and `/*nLocalHostNonceIn=*/ 1`

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19998#issuecomment-703090033).

jonatack reviewed (commented):
ACK. The new unit tests are green for me locally. Not sure what's up with the travis linter.

hebasto commented:
Updated 5fc8f155549b1ad39aeef1c3220ebd6026d7f845 -> 3989fcfb6713b074c6c6be143db3aaa33e50525a ([pr19998.10](https://github.com/hebasto/bitcoin/commits/pr19998.10) -> [pr19998.11](https://github.com/hebasto/bitcoin/commits/pr19998.11)):

- addressed all of the recent @jonatack's comments
- rebased to avoid a weird CI [error message](https://travis-ci.org/github/bitcoin/bitcoin/jobs/732481553)

jonatack reviewed (approved):
ACK 3989fcfb6713b074c6c6be143db3aaa33e50525a

hebasto commented on src/net.h:
```diff
@@ -954,6 +961,16 @@ class CNode
+     * Get network the peer connected through.
+     *
+     * CNetAddr::GetNetClass() is used excluding inbound connections that are
+     * explicitly tagged as connections from the Tor onion service(s).
```
It's for sure that my comment could be improved, but I don't think that it is a correct assumption that it's responsibility of `CNetAddr` class to detect the network a peer is connected through. I mean "CNetAddr::GetNetClass() does not currently detect" part.

jonatack replied:
Yes. That seems like a worthwhile clarification. Maybe:

`Returns Network::NET_ONION for *inbound* onion connections, and CNetAddr::GetNetClass() otherwise. The latter cannot be used directly because it's not the responsability of the CNetAddr class to know the actual network a peer is connected through.`

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19998#issuecomment-703097616).

hebasto commented:
Updated 3989fcfb6713b074c6c6be143db3aaa33e50525a -> 3984b78cd7f49e409377f2175a56e8e4bd71d1d8 ([pr19998.11](https://github.com/hebasto/bitcoin/commits/pr19998.11) -> [pr19998.12](https://github.com/hebasto/bitcoin/commits/pr19998.12), [diff](https://github.com/hebasto/bitcoin/compare/pr19998.11..pr19998.12)):

- addressed @jonatack's [comment](https://github.com/bitcoin/bitcoin/pull/19998#discussion_r499143917)

jonatack commented:
Thanks for your patience with the suggestions.

re-ACK 3984b78cd7f49e409377f2175a56e8e4bd71d1d8 per `git diff 3989fcf 3984b78c`

laanwj commented:
Code review ACK 3984b78cd7f49e409377f2175a56e8e4bd71d1d8
Thanks for addressing my suggestion.
